src/Transform/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -329,56 +311,63 @@ void DimAnalysis::visitDim(
   // possible.
   Operation *op = tensor.getDefiningOp();
 
-  // UnaryOp
-  if (isa<ONNXCastOp>(op) || isa<ONNXPowOp>(op) || isa<ONNXReluOp>(op) ||
-      isa<ONNXReciprocalOp>(op) || isa<ONNXTanhOp>(op) ||
-      isa<ONNXSigmoidOp>(op) || isa<ONNXSoftmaxOp>(op) || isa<ONNXSqrtOp>(op)) {
-    exploreSameInputDimsUnaryOp(dim, op, sameDims);
-    return;
-  }
-
-  // BinaryOp
-  if (isa<ONNXAddOp>(op) || isa<ONNXDivOp>(op) || isa<ONNXMulOp>(op) ||
-      isa<ONNXSubOp>(op)) {
-    exploreSameInputDimsBinaryOp(dim, op, sameDims);
-    // If we know by this analysis that two unknown dims at the same index are
-    // the same, then the output dim must be the same too.
-    Value A = op->getOperands()[0];
-    Value B = op->getOperands()[1];
-    Type aType = A.getType();
-    Type bType = B.getType();
-    uint64_t aRank = onnx_mlir::getRank(aType);
-    uint64_t bRank = onnx_mlir::getRank(bType);
-    uint64_t maxRank = std::max(aRank, bRank);
-    if ((aRank != 0) && (bRank != 0)) {
-      ArrayRef<int64_t> aShape = onnx_mlir::getShape(aType);
-      ArrayRef<int64_t> bShape = onnx_mlir::getShape(bType);
-      // aDim == bDim (unknown), there is no broadcasting and aDim == outputDim.
-      int64_t aDimIndex = dimIndex - (maxRank - aRank);
-      int64_t bDimIndex = dimIndex - (maxRank - bRank);
-      if ((aDimIndex >= 0) && (bDimIndex >= 0) &&
-          ShapedType::isDynamic(aShape[aDimIndex]) &&
-          ShapedType::isDynamic(bShape[bDimIndex]) &&
-          onnx_mlir::DimAnalysis::sameUnknownDim(A, aDimIndex, B, bDimIndex))
-        sameDims.insert(onnx_mlir::DimAnalysis::DimT(A, aDimIndex));
+  ////////////////////////////////////////////////////
+  // Special cases.
+
+  // BinaryOp (alphabetical list, same as under ShapeHelper
+  // ONNXBroadcastOpShapeHelper).
+  // clang-format off
+  if (isa<ONNXAddOp>(op) || 
+      isa<ONNXAndOp>(op) ||
+      isa<ONNXBitShiftOp>(op) ||
+      isa<ONNXBitwiseAndOp>(op) ||
+      isa<ONNXBitwiseOrOp>(op) ||
+      isa<ONNXBitwiseXorOp>(op) ||
+      isa<ONNXBitShiftOp>(op) ||
+      isa<ONNXDivOp>(op) || 
+      isa<ONNXEqualOp>(op) ||
+      isa<ONNXGreaterOp>(op) ||
+      isa<ONNXGreaterOrEqualOp>(op) ||
+      isa<ONNXLessOp>(op) ||
+      isa<ONNXLessOrEqualOp>(op) ||
+      isa<ONNXMeanOp>(op) ||
+      isa<ONNXMinOp>(op) ||
+      isa<ONNXModOp>(op) ||
+      isa<ONNXMulOp>(op) ||
+      isa<ONNXOrOp>(op) ||
+      isa<ONNXPowOp>(op) ||
+      isa<ONNXSubOp>(op) ||
+      isa<ONNXSumOp>(op) ||
+      isa<ONNXWhereOp>(op) ||
+      isa<ONNXXorOp>(op)) {
+    // c lang-format on
+    if (exploreSameInputDims(dim, op, sameDims)) {
```

Comment:
It seems the logic is different from the old code.

In the old code, the following code is executed even if `exploreSameInputDims` failed.

Reply:
@tungld so you are telling me the test can be performed regardless of if we successfully ran the check? Note that is is known to work unconditionally for all these ops, its more like unsupported ops that would fail. But I can remove the test if you feel more comfortable with that.

Reply:
Changed the `if (explore...)` to `assert(explore...)` since they are guaranteed to work for these ops. Prefer checking my assumptions that just hoping it should work.